drivers/i2c/i2c_omap.c

```diff
@@ -0,0 +1,815 @@
+#define I2C_OMAP_BUF_RXFIF_CLR BIT(14) /* RX FIFO Clear */
+#define I2C_OMAP_BUF_TXFIF_CLR BIT(6)  /* TX FIFO Clear */
+
+/* I2C Status Register (I2C_OMAP_STAT): */
+#define I2C_OMAP_STAT_XDR  BIT(14) /* TX Buffer draining */
+#define I2C_OMAP_STAT_RDR  BIT(13) /* RX Buffer draining */
+#define I2C_OMAP_STAT_BB   BIT(12) /* Bus busy */
+#define I2C_OMAP_STAT_ROVR BIT(11) /* Receive overrun */
+#define I2C_OMAP_STAT_XUDF BIT(10) /* Transmit underflow */
+#define I2C_OMAP_STAT_AAS  BIT(9)  /* Address as target */
+#define I2C_OMAP_STAT_XRDY BIT(4)  /* Transmit data ready */
+#define I2C_OMAP_STAT_RRDY BIT(3)  /* Receive data ready */
+#define I2C_OMAP_STAT_ARDY BIT(2)  /* Register access ready */
+#define I2C_OMAP_STAT_NACK BIT(1)  /* No ack interrupt enable */
+#define I2C_OMAP_STAT_AL   BIT(0)  /* Arbitration lost */
+
+/* I2C System Test Register (I2C_OMAP_SYSTEST): */
+#define I2C_OMAP_SYSTEST_ST_EN       BIT(15)   /* System test enable */
+#define I2C_OMAP_SYSTEST_FREE        BIT(14)   /* Free running mode */
+#define I2C_OMAP_SYSTEST_TMODE_MASK  (3 << 12) /* Test mode select mask */
+#define I2C_OMAP_SYSTEST_TMODE_SHIFT (12)      /* Test mode select shift */
+
+/* Functional mode */
+#define I2C_OMAP_SYSTEST_SCL_I_FUNC BIT(8) /* SCL line input value */
+#define I2C_OMAP_SYSTEST_SDA_I_FUNC BIT(6) /* SDA line input value */
+
+/* SDA/SCL IO mode */
+#define I2C_OMAP_SYSTEST_SCL_I BIT(3) /* SCL line sense in */
+#define I2C_OMAP_SYSTEST_SCL_O BIT(2) /* SCL line drive out */
+#define I2C_OMAP_SYSTEST_SDA_I BIT(1) /* SDA line sense in */
+#define I2C_OMAP_SYSTEST_SDA_O BIT(0) /* SDA line drive out */
+
+typedef void (*init_func_t)(const struct device *dev);
+#define DEV_CFG(dev)      ((const struct i2c_omap_cfg *)(dev)->config)
+#define DEV_DATA(dev)     ((struct i2c_omap_data *)(dev)->data)
+#define DEV_I2C_BASE(dev) ((i2c_omap_regs_t *)DEVICE_MMIO_NAMED_GET(dev, base))
+
+struct i2c_omap_cfg {
+	DEVICE_MMIO_NAMED_ROM(base);
+	uint32_t irq;
+	uint32_t speed;
+	init_func_t init_func;
+};
+
+enum i2c_omap_speed {
+	I2C_OMAP_SPEED_STANDARD,
+	I2C_OMAP_SPEED_FAST,
+	I2C_OMAP_SPEED_FAST_PLUS,
+};
+
+struct i2c_omap_speed_config {
+	uint32_t pscstate;
+	uint32_t scllstate;
+	uint32_t sclhstate;
+};
+
+struct i2c_omap_data {
+	DEVICE_MMIO_NAMED_RAM(base);
+	enum i2c_omap_speed speed;
+	struct i2c_omap_speed_config speed_config;
+	struct i2c_msg current_msg;
+	struct k_sem cmd_complete;
+	int cmd_err;
+	bool receiver;
+	bool bb_valid;
+};
+
+/**
+ * @brief Acknowledge the I2C status by writing to the I2C_OMAP_STAT register.
+ *
+ * This function is used to acknowledge the I2C status by writing the provided
+ * status value to the I2C_OMAP_STAT register. It is an inline function that is
+ * specific to the OMAP I2C driver.
+ *
+ * @param dev Pointer to the device structure.
+ * @param stat The status value to be written to the I2C_OMAP_STAT register.
+ */
+static inline void i2c_omap_ack_stat(const struct device *dev, uint16_t stat)
```

Comment:
Everywhere this function is used, `i2c_base_addr` is already known, you could just pass that into this function.

Or even better, since that would make this function a one liner `i2c_base_addr->STAT = stat;`, just do that directly instead of calling this function below.

Reply:
Got it.
I will update it!